semantic-core/src/Data/Core/Parser.hs

```diff
@@ -48,65 +50,79 @@ core :: (TokenParsing m, Monad m) => m (Term Core User)
 core = expr
 
 expr :: (TokenParsing m, Monad m) => m (Term Core User)
-expr = atom `chainl1` go where
-  go = choice [ (Core....) <$ dot
-              , (Core.$$)  <$ notFollowedBy dot
-              ]
+expr = ifthenelse <|> lambda <|> rec <|> load <|> assign
+
+assign :: (TokenParsing m, Monad m) => m (Term Core User)
+assign = application <**> (flip (Core..=) <$ symbolic '=' <*> application <|> pure id) <?> "assignment"
+
+application :: (TokenParsing m, Monad m) => m (Term Core User)
+application = projection `chainl1` (pure (Core.$$))
+
+projection :: (TokenParsing m, Monad m) => m (Term Core User)
+projection = foldl' (Core....) <$> atom <*> many (namedValue <$ dot <*> name)
 
 atom :: (TokenParsing m, Monad m) => m (Term Core User)
 atom = choice
   [ comp
-  , ifthenelse
-  , edge
   , lit
   , ident
-  , assign
   , parens expr
   ]
 
 comp :: (TokenParsing m, Monad m) => m (Term Core User)
-comp = braces (Core.block <$> sepEndByNonEmpty expr semi) <?> "compound statement"
+comp = braces (Core.do' <$> sepEndByNonEmpty statement semi) <?> "compound statement"
+
+statement :: (TokenParsing m, Monad m) => m (Maybe (Named User) Core.:<- Term Core User)
+statement
+  =   try ((Core.:<-) . Just <$> name <* symbol "<-" <*> expr)
```

Comment:
Nitpick: I would give `(Core.:<-)` its own prefix name here, as that particular punctuation cluster is tough to read when prefixed with `Core` (or just import it unqualified)

Reply:
Fixed in 78bd32e0b.